---
title: "RNA-seq Analysis"
author: "Giovani Genesi, MSc / Phillips Lab (UPenn)"
date: "`r Sys.Date()`"
output: pdf_document
---

```{r Load packages, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
# Set global options
knitr::opts_chunk$set(echo = FALSE, 
                      warning = FALSE, 
                      message = FALSE, 
                      include = TRUE)

# This code will check if required packages are installed. 
# If not, this code will install these packages.
if (!require("BiocManager", quietly = TRUE)){
  install.packages("BiocManager")
}

# Load required packages
if(!require("ggplot2", quietly = TRUE)){install.packages("ggplot2")}
library(ggplot2)
if(!require("ggrepel", quietly = TRUE)){install.packages("ggrepel")}
library(ggrepel)
if(!require("dplyr", quietly = TRUE)){install.packages("dplyr")}
library(dplyr)
if(!require("tidyr", quietly = TRUE)){install.packages("tidyr")}
library(tidyr)
if(!require("tidyverse", quietly = TRUE)){install.packages("tidyverse")}
library(tidyverse)
if(!require("stringr", quietly = TRUE)){install.packages("stringr")}
library(stringr)
if(!require("DESeq2", quietly = TRUE)){BiocManager::install("DESeq2")}
library(DESeq2)
# library(tidytext)
# library(ggrepel)
```

### Change for each project
```{r Setup Working Directory}
### CHANGE FOR EACH PROJECT
# Specify working directory
work_dir = "C:/Users/thephillipslab/Documents/Projects/RNAseq_Brien_paper/"
knitr::opts_knit$set(root.dir = work_dir)


# Output settings
project_name = "RNAseq_Brien"

# Set path to importable data folder 
import_dir = file.path(work_dir, "Processing and Analysis/Importable Data/")

# Set path to importable data folder 
metadata_dir = file.path(work_dir, "Original Data and Metadata/Metadata/")

# Set path to results folder 
results_dir = file.path(work_dir, "Processing and Analysis/Analysis Data/")

```


```{r Prepare function to make volcano plot}
# Function will take the results data frame from DESeq2 as input
makeVolcanoPlot <- function(results_df, lfc.threshold = 1, ylim = NULL){
  volcanoPlot <- ggplot(data=results_df, 
                        aes(x=log2FoldChange, 
                            y=-log10(padj),
                            color = effect, 
                            label = Label)) + 
  geom_point() + 
  theme_classic() +
  # Add axes labels
  ylab(expression(bold(bolditalic(-Log[10]) ~ "FDR"))) +
  xlab(expression(bold(bolditalic(Log[2]) ~ "Fold-Change"))) +
  # Add plot tile based on which comparison was made
  ggtitle(comparison) +
  # Add vertical lines for log2FoldChange thresholds, and one horizontal line for the p-value threshold 
  geom_vline(xintercept = -lfc.threshold, lty = "dashed") +
  geom_vline(xintercept = lfc.threshold, lty = "dashed") +
  geom_hline(yintercept = -log10(alpha), lty = "dashed") +
  scale_x_continuous(n.breaks = 7, 
                     ) +
  scale_y_continuous(limits = ylim) +
  # Modify plotting theme
  theme(legend.title = element_blank(),
        plot.title = element_text(hjust = 0.5),
        panel.border = element_rect(color = "black", fill = NA),
        panel.grid.minor = element_blank()) +
  # Change point color 
  scale_color_manual(values=c("Downregulated" = "navyblue",
                              "NS" = "gray", 
                              "Upregulated" = "red")) + 
  # Organize labels in the plot
  geom_text_repel(nudge_x = 0.5 * results_df$log2FoldChange)  
  
  return(volcanoPlot)
}

```

```{r Prepare function to make volcano plot for all genes}
# Function will take two data frames as input
makeVolcanoPlot2 <- function(results_allgenes, 
                             results_subset, 
                             lfc.threshold = 1, 
                             ylim = NULL){
  
  volcanoPlot <- ggplot(data=results_allgenes, 
                        aes(x=log2FoldChange, 
                            y=-log10(padj),
                            fill = effect,
                            color = effect,
                            label = Label)) +
    
    geom_point(alpha = 0.5, shape = 21) +
    geom_point(data = results_subset, 
               aes(x=log2FoldChange, 
                   y=-log10(padj), 
                   fill = effect), 
               shape = 21, 
               size = 2, 
               color = "black") +
    
    theme_classic() +
    # Add axes labels
    ylab(expression(bold(bolditalic(-Log[10]) ~ "FDR"))) +
    xlab(expression(bold(bolditalic(Log[2]) ~ "Fold-Change"))) +
    # Add plot tile based on which comparison was made
    ggtitle(comparison) +
    # Add vertical lines for log2FoldChange thresholds, and one horizontal line for the p-value threshold 
    geom_vline(xintercept = -lfc.threshold, lty = "dashed") +
    geom_vline(xintercept = lfc.threshold, lty = "dashed") +
    geom_hline(yintercept = -log10(alpha), lty = "dashed") +
    scale_x_continuous(n.breaks = 7, 
                       ) +
    scale_y_continuous(limits = ylim) +
    # Modify plotting theme
    theme(legend.title = element_blank(),
          plot.title = element_text(hjust = 0.5),
          panel.border = element_rect(color = "black", fill = NA),
          panel.grid.minor = element_blank()) +
    # Change point color 
    scale_fill_manual(values=c("Downregulated" = "blue",
                                "NS" = "gray", 
                                "Upregulated" = "red")) + 
    scale_color_manual(values=c("Downregulated" = "blue",
                                "NS" = "gray", 
                                "Upregulated" = "red")) + 
    guides(color = "none")  + 
    # Organize labels in the plot
    geom_text_repel(color = "black", nudge_x = 0.5 * results_allgenes$log2FoldChange) 
    
  return(volcanoPlot)
}

```

```{r Load PcB target genes}
# This code loads the information about the PcB targets in DIPGXIII and keeps
# gene names
PcB_genes <- read.csv(file.path(metadata_dir, 
                                "DIPGXIII_PcG_consensus_target_genes.csv"), 
                      header = T)

PcB_gene_names = PcB_genes["Gene.Name"]

# Store PcB target genes
saveRDS(PcB_genes, file = file.path(import_dir, "PcB_genes.rds"))
saveRDS(PcB_gene_names, file = file.path(import_dir, "PcB_gene_names.rds"))
```



```{r Load Gene Expression Data}
# Load raw counts
counts <- read.delim(file.path(import_dir,
                               "RNAseq_counts_Brien_paper.txt"), 
                     header = T, 
                     row.names = 1)

# counts %>% head() # check the loaded count matrix
# 
# colnames(counts) # make sure columns correctly correspond to samples' names

```


```{r Prepare Metadata}
# Load sample metadata
sample_metadata <- read.csv(file.path(metadata_dir, "sample_metadata.csv"), 
                            header = T)

sample_metadata$genotype = sub("-", "_", sample_metadata$genotype) %>%
  as.factor()
sample_metadata$guide = as.factor(sample_metadata$guide)
sample_metadata$target = as.factor(sample_metadata$target)

# Add harvest metadata
sample_metadata <- sample_metadata %>%
  mutate(sample.ID_A = "A",
         sample.ID_B = "B") %>%
  gather(key = "type", value = "harvest", sample.ID_A, sample.ID_B) %>%
  select(-type) %>%
  arrange(sample.ID)

sample_metadata$harvest = as.factor(sample_metadata$harvest)

# Add condition pooling timepoints
sample_metadata$condition = paste(sample_metadata$guide,
                                  sample_metadata$genotype,
                                  sep = "_") %>% as.factor()

# Add condition pooling gRNAs
sample_metadata$treatment = paste(sample_metadata$target,
                                  sample_metadata$genotype,
                                  sep = "_") %>% as.factor()

# Add condition pooling gRNAs for each harvest
sample_metadata$group = paste(sample_metadata$target,
                                  sample_metadata$genotype,
                                  sample_metadata$harvest,
                                  sep = "_") %>% as.factor()

# Add sample name
sample_metadata$sample.name = paste(sample_metadata$guide,
                                  sample_metadata$genotype,
                                  sample_metadata$replicate,
                                  sample_metadata$harvest,
                                  sep = "_") %>% as.factor()

# Group sample ID and harvest to match colnames of count matrix
sample_metadata$ID.harvests = paste0("X", 
                                     sample_metadata$sample.ID,
                                     sample_metadata$harvest)

# Match sample IDs to colnames in the counts matrix
sample_metadata <- sample_metadata %>%
  mutate(colnames = sapply(ID.harvests, function(harvest) {
    matching_colname <- colnames(counts)[str_detect(colnames(counts), harvest)]
    if(length(matching_colname) > 0) {
      return(matching_colname[1]) # Return the first matching value
    } else {
      return(NA) # If no match is found, return NA
    }
  }))

# Make the rownames equal to sample names
rownames(sample_metadata) = sample_metadata$sample.name

# Remove samples that did not match
sample_metadata = filter(sample_metadata, !is.na(sample_metadata$colnames))

# Store edited sample metadata
saveRDS(sample_metadata, file = file.path(import_dir, "sample_metadata.rds"))
```

```{r Prepare counts matrix for data analysis}
# This code will reorder the columns in the count matrix to match the sample
# order in the sample metadata
counts = counts[, sample_metadata$colnames]

# The following code will replace the colnames in the count matrix 
# with appropriate sample names 

## 1. Create a named vector for replacement
replacement_vector <- setNames(sample_metadata$sample.name, 
                               sample_metadata$colnames)

## 2. Replace the column names in counts
new_colnames <- replacement_vector[colnames(counts)]
colnames(counts) <- new_colnames

## 3. Check if colnames of countr matrix match rownames of metadata
# all(rownames(sample_metadata) == colnames(counts))

# Make the control the reference level for each factor
sample_metadata$guide = relevel(sample_metadata$guide, 
                                "LUC")
print("Levels of gRNA factor:")
levels(sample_metadata$guide)

sample_metadata$target = relevel(sample_metadata$target, 
                                 "LUC")
print("Levels of gene target factor:")
levels(sample_metadata$target)

sample_metadata$genotype = relevel(sample_metadata$genotype,
                                   "K27M+")
print("Levels of genotype factor:")
levels(sample_metadata$genotype)

sample_metadata$condition = relevel(sample_metadata$condition, 
                                    "LUC_K27M+")
print("Levels of condition factor:")
levels(sample_metadata$condition)

sample_metadata$treatment = relevel(sample_metadata$treatment, 
                                    "LUC_K27M+")
print("Levels of treatment factor:")
levels(sample_metadata$treatment)

sample_metadata$group = relevel(sample_metadata$group, 
                                "LUC_K27M+_A")
print("Levels of group factor:")
levels(sample_metadata$group)

# Save the edited count matrix
saveRDS(counts, file = file.path(import_dir, "counts.rds"))
```

```{r QC - Check counts per sample}
# Set up colours
cols = factor(sample_metadata$target) %>% as.numeric()

# Convert count matrix to df for plotting with ggplot
df_counts <- data.frame(
  Sample = colnames(counts),
  Counts = colSums(counts)
)

# Ensure Sample is a factor with levels in the desired order
df_counts$Sample <- factor(df_counts$Sample, levels = df_counts$Sample)

# Make barplot
bp <- ggplot(df_counts, aes(x = Sample, y = Counts, fill = Sample)) +
  geom_bar(stat = "identity", show.legend = FALSE, color = "black") +
  scale_fill_manual(values = cols) +
  labs(title = "Reads per sample") +
  xlab("Raw counts") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 8),
        axis.text.y = element_text(size = 8),
        plot.title = element_text(hjust = 0.5))

# Export barplot
bp_path = file.path(results_dir,
                    paste0("counts_per_index.pdf"))

pdf(file = bp_path, width = 10, height = 5)
print(bp)
dev.off()
```

```{r Create DESeqDataSet}
# Create DESeqDataSet
dataSet <- DESeqDataSetFromMatrix(countData = counts, # count matrix
                                  colData = sample_metadata , # samples&grouping
                                  design = ~ harvest + condition
                                  )

# Pre-filtering genes with low total counts
nrow(dataSet)

smallestGroupSize <- 2
keep <- rowSums(counts(dataSet) >= 10) >= smallestGroupSize
dataSet <- dataSet[keep,]
nrow(dataSet)

# Create DESeq object
DESeq <- DESeq(dataSet,
               #test="LRT", reduced=~sgRNA + treatment
               )

DESeq2::plotDispEsts(DESeq)



```

```{r Statistical test}
# FDR cut-off
alpha = 0.1

# LFC cut-off
lfc = 0.585

# List comparisons for statistical testing

## Pooling timepoints
# output_dir = "Pooling Timepoints"
# contrast_by = "condition"
#
# controls = c("LUC_K27M+", "LUC_K27M_KO")
#
# conditions = c("LUC_K27M_KO", "CBX4.1_K27M_KO", "CBX4.2_K27M_KO", "CBX2.1_K27M_KO",
#                "CBX2.3_K27M_KO","EZH2.3_K27M_KO","EZH2.5_K27M_KO","CBX8.1_K27M_KO",
#                "CBX8.2_K27M_KO","BMI1.1_K27M_KO","BMI1.3_K27M_KO","LUC_K27M+",
#                "CBX4.1_K27M+","CBX4.2_K27M+","CBX2.1_K27M+","CBX2.3_K27M+",
#                "EZH2.3_K27M+","EZH2.5_K27M+","CBX8.1_K27M+","CBX8.2_K27M+",
#                "BMI1.1_K27M+","BMI1.3_K27M+")

## Pooling timepoints, taking timepoint into account in the design matrix
output_dir = "Harvest plus Condition"
contrast_by = "condition"

controls = c("LUC_K27M+", "LUC_K27M_KO")

conditions = c("LUC_K27M_KO", "CBX4.1_K27M_KO", "CBX4.2_K27M_KO", "CBX2.1_K27M_KO",
               "CBX2.3_K27M_KO","EZH2.3_K27M_KO","EZH2.5_K27M_KO","CBX8.1_K27M_KO",
               "CBX8.2_K27M_KO","BMI1.1_K27M_KO","BMI1.3_K27M_KO","LUC_K27M+",
               "CBX4.1_K27M+","CBX4.2_K27M+","CBX2.1_K27M+","CBX2.3_K27M+",
               "EZH2.3_K27M+","EZH2.5_K27M+","CBX8.1_K27M+","CBX8.2_K27M+",
               "BMI1.1_K27M+","BMI1.3_K27M+")

## Pooling gRNAs
# output_dir = "Pooled guides"
# contrast_by = "group"
#
# controls = c("LUC_K27M+_A", "LUC_K27M_KO_A","LUC_K27M+_B", "LUC_K27M_KO_B")
#
# conditions = c("LUC_K27M_KO_A","LUC_K27M_KO_B","CBX4_K27M_KO_A","CBX4_K27M_KO_B",
#                "CBX2_K27M_KO_A","CBX2_K27M_KO_B","EZH2_K27M_KO_A",
#                "EZH2_K27M_KO_B","CBX8_K27M_KO_A","CBX8_K27M_KO_B","BMI1_K27M_KO_A",
#                "BMI1_K27M_KO_B","LUC_K27M+_A","LUC_K27M+_B","CBX4_K27M+_A",
#                "CBX4_K27M+_B","CBX2_K27M+_A","CBX2_K27M+_B","EZH2_K27M+_A",
#                "EZH2_K27M+_B","CBX8_K27M+_A","CBX8_K27M+_B","BMI1_K27M+_A",
#                "BMI1_K27M+_B")

## Pooling gRNAs, taking timepoint into account
# output_dir = "Harvest plus Treatment"
# contrast_by = "treatment"
#
# controls = c("LUC_K27M+", "LUC_K27M_KO")
#
# conditions = c("LUC_K27M_KO","CBX4_K27M_KO","CBX2_K27M_KO","EZH2_K27M_KO",
#                "CBX8_K27M_KO","BMI1_K27M_KO","LUC_K27M+","CBX4_K27M+",
#                "CBX2_K27M+","EZH2_K27M+","CBX8_K27M+","BMI1_K27M+")


output.dir = file.path(results_dir, output_dir)
if(!dir.exists(output.dir)){
      dir.create(output.dir)
    }

# This code loops through each possible treatment vs control combination,
# performs statistical testing, and outputs table of results
for(control in 1:length(controls)){
  for(condition in 1:length(conditions)){
    if(conditions[condition] == controls[control]){
      next
    }
    # Select an specific comparison
    comparisons = paste(conditions[condition], "vs", controls[control], sep = "_")

    # Contrast
    contrast = c(contrast_by,conditions[condition],controls[control])
    print(contrast)
    # Building the results table
    results <- results(DESeq,
                       contrast = contrast,
                       alpha = alpha)

    comp_dir = file.path(results_dir, output_dir, comparisons)
    if(!dir.exists(comp_dir)){
      dir.create(comp_dir)
    }

    # Export results table
    results_df <- as.data.frame(results) %>% arrange(padj)

    results_df = results_df %>% rownames_to_column(var = "Gene")

    results_df$effect = "NS"
    results_df$effect[results_df$padj < alpha &
                        results_df$log2FoldChange <= -lfc] = "Downregulated"
    results_df$effect[results_df$padj < alpha &
                        results_df$log2FoldChange >= lfc] = "Upregulated"

    # Prepare data for volcano plot
    results_PcB = filter(results_df, Gene %in% PcB_gene_names$Gene.Name)
    
    results_PcB$Label = NA
    results_PcB$Label[results_PcB$effect != "NS"] = results_PcB$Gene[results_PcB$effect != "NS"]
    
    is_labeled = na.omit(results_PcB$Label)

    results_df$Label = NA
    results_df$Label[results_df$Gene %in% is_labeled] = results_df$Gene[results_df$Gene %in% is_labeled]
    
    # Make volcano plot and export
    comparison = gsub("_", " ", comparisons)
    vp = makeVolcanoPlot(results_PcB, lfc.threshold = lfc)
    vp2 = makeVolcanoPlot2(results_df, results_PcB, lfc.threshold = lfc)

    # Export volcano plot
    vp_path = file.path(comp_dir,
                        paste0("volcano_PcB_targets_", comparisons, ".pdf"))

    pdf(file = vp_path, width = 6.5, height = 5)
    print(vp)
    dev.off()
    
    # Export volcano plot for all genes
    vp_path = file.path(comp_dir,
                        paste0("volcano_all_genes_", comparisons, ".pdf"))

    pdf(file = vp_path, width = 6.5, height = 5)
    print(vp2)
    dev.off()

    # Export data frames with results
    df_path = file.path(comp_dir,
                        paste0("results_", comparisons, ".csv"))

    write.csv(results_df, file = df_path, quote = FALSE)
    
    df_path = file.path(comp_dir,
                        paste0("PcB_targets_", comparisons, ".csv"))
    write.csv(results_PcB, file = df_path, quote = FALSE)
    
  }
}

contrast_by = tolower(gsub(" ", "_", output_dir))
rds_name = paste0("DESeq_",contrast_by,".rds")

saveRDS(DESeq,
        file = file.path(results_dir, rds_name))
```




```{r Gene Expression Heatmap}

# if (!require("pheatmap")) {
#   install.packages("pheatmap", dependencies = TRUE)
#   }
# library(pheatmap)
# library(RColorBrewer)
# 
# rlogData <- rlog(dataSet, blind = FALSE)
# rlogData %>% assay() %>% head()
# 
# rlogCounts = assay(rlogData) %>% as.data.frame()
# rlogCounts = rlogCounts %>% mutate(label = rownames(rlogCounts))
# 
# n_all = nrow(rlogCounts)
# 
# hm = pheatmap(rlogCounts,
#          #color = c("black","white","darkgoldenrod1"),
#          color = colorRampPalette(rev(brewer.pal(n = 7, name =
#                                                    "RdBu")))(100),
#          scale="row",
#          clustering_distance_rows = "euclidean",
#          clustering_distance_cols = "euclidean",
#          clustering_method = "complete",
#          show_rownames = F,
#          main = paste0("Differentially expressed genes\n N = ", n_all),
#          angle_col = 45,
#          labels_col = colnames(rlogData)
#          )
# 
# hm_path = file.path(comp_dir,
#                         paste0("heatmap_all_genes.pdf"))
# 
# pdf(file = hm_path, width = 10, height = 5)
# print(hm)
# dev.off()
# 
# hm

```


```{r Gene Expression Heatmap - PcB targets}

# rlogCounts = assay(rlogData) %>% as.data.frame()
# 
# rlogCounts_PcB = filter(rlogCounts, rownames(rlogCounts) %in% PcB_gene_names$Gene.Name)
# 
# # rlogCounts_K27M = rlogCounts %>% select(-contains("K27M_KO"))
# # rlogCounts_K27M_PcB = rlogCounts_PcB %>% select(-contains("K27M_KO"))
# # 
# # rlogCounts_K27M_KO = rlogCounts %>% select(-contains("K27M+"))
# # rlogCounts_K27M_KO_PcB = rlogCounts_PcB %>% select(-contains("K27M+"))
# 
# 
# n_PcB = nrow(rlogCounts_PcB)
# labels_cols = colnames(rlogCounts)
# # labels_K27M = colnames(rlogCounts_K27M)
# # labels_K27M_KO = colnames(rlogCounts_K27M_KO)
# 
# hm = pheatmap(rlogCounts_K27M_KO,
#          color = colorRampPalette(rev(brewer.pal(n = 7, name =
#                                                    "RdBu")))(100),
#          scale="row",
#          clustering_distance_rows = "euclidean",
#          clustering_distance_cols = "euclidean",
#          clustering_method = "complete",
#          show_rownames = F,
#          main = paste0("Differentially expressed genes (PcB targets)\n N = ", n_PcB),
#          angle_col = 45,
#          labels_col = labels_K27M_KO
#          )
# 
# hm_path = file.path(comp_dir,
#                         paste0("heatmap_PcB_targets.pdf"))
# 
# pdf(file = hm_path, width = 10, height = 5)
# print(hm)
# dev.off()
# 
# hm
```



```{r Prepare data for Q-Q plot}
# Pattern matching for each condition
condition1 = "K27M_KO"
condition2 = "K27M+"
targets = c("CBX4.1", "CBX4.2", "CBX2.1","CBX2.3","EZH2.3","EZH2.5",
            "CBX8.1","CBX8.2","BMI1.1","BMI1.3")

subfolder = "Harvest plus Condition"

output_dir = file.path(results_dir, "Genotype comparisons")

if(!dir.exists(output_dir)){
  dir.create(output_dir)
}

search_dir = file.path(results_dir, subfolder)
list_dir = list.dirs(search_dir)

for(i in 1:length(targets)){
  t = paste0(targets[i], "_", condition1, "_vs_LUC_", condition1) # K27M KO
  t2 = paste0(targets[i], "_", condition2, "_vs_LUC_", condition2) # K27M+

  comp = gsub("_", " ", t) # K27M KO
  comp2 = gsub("_", " ", t2) # K27M+

  compare = paste0(targets[i], "_", condition2, "_vs_", condition1)
  comp_dir = file.path(output_dir, compare)

  print(paste0("Making comparison: ", compare))

  if(!dir.exists(comp_dir)){
    dir.create(comp_dir)
  }

  t = gsub("([.|()\\^{}+$*?]|\\[|\\])", "\\\\\\1", t) # K27M KO
  t2 = gsub("([.|()\\^{}+$*?]|\\[|\\])", "\\\\\\1", t2) # K27M+

  f1 = list_dir[grepl(t, list_dir)]  # K27M KO
  f2 = list_dir[grepl(t2, list_dir)]  # K27M+

  csv1 = file.path(f1, paste0("PcB_targets_", basename(f1), ".csv")) # K27M KO
  csv2 = file.path(f2, paste0("PcB_targets_", basename(f2), ".csv")) # K27M+

  tab1 = read.csv(csv1)[,-1] # K27M KO
  tab2 = read.csv(csv2)[,-1] # K27M+

  tab_merge = merge(tab1, tab2, by = "Gene")

  tab_merge$Label = NA
  tab_merge$Label[!is.na(tab_merge$Label.x)] = tab_merge$Gene[!is.na(tab_merge$Label.x)]
  tab_merge$Label[!is.na(tab_merge$Label.y)] = tab_merge$Gene[!is.na(tab_merge$Label.y)]

  limits = c(min(range(tab_merge$log2FoldChange.x)), # K27M KO
             max(range(tab_merge$log2FoldChange.y))) %>%  # K27M+
    rev()

  tab_merge = filter(tab_merge, !is.na(tab_merge$Label))

  # x = K27M KO
  # y = K27M+

  # Create the scatter plot
  qqp <- ggplot(tab_merge, aes(x = log2FoldChange.x,
                      y = log2FoldChange.y
                      )
  ) +
    geom_point(size = 2, fill = "black", shape = 21) +
    geom_point(data = filter(tab_merge, log2FoldChange.x - log2FoldChange.y >= 1),
               size = 2, fill = "blue", color = "blue", shape = 21) +
    geom_point(data = filter(tab_merge, log2FoldChange.x - log2FoldChange.y <= -1),
               size = 2, fill = "red", color = "red", shape = 21) +
    # Add reference lines
    geom_abline(intercept = 0, slope = 1, color = "black") +
    geom_abline(intercept = 1, slope = 1, linetype = "dashed", color = "red") +
    geom_abline(intercept = -1, slope = 1, linetype = "dashed", color = "blue") +
    labs(x = paste0("Log2 Fold-Change (", comp, ")"),
         y = paste0("Log2 Fold-Change (", comp2, ")")) +
    scale_x_continuous(limits = c(-limits[1],-limits[2])) +
    scale_y_continuous(limits = c(-limits[1],-limits[2])) +
    geom_text_repel(aes(label = Label),
                    nudge_y = tab_merge$log2FoldChange.y / 2,
                    nudge_x = tab_merge$log2FoldChange.x / 1,
                    max.overlaps = 10
    ) +
    theme_classic() +
    guides(color = "none", alpha = "none")

  # Export q-q plot
    qqp_path = file.path(comp_dir,
                        paste0("Q-Q-plot_", compare, ".pdf"))

    pdf(file = qqp_path, width = 5, height = 5)
    print(qqp)
    dev.off()


}



```


```{r Make scaled volcano plot}
condition = "K27M+"
ref_targets = c("CBX4.1", "CBX4.2")

targets = c("CBX4.1", "CBX4.2", "CBX2.1", "CBX2.3", "CBX8.1", "CBX8.2")
subfolder = "Harvest plus Condition"

search_dir = file.path(results_dir, subfolder)
list_dir = list.dirs(search_dir)

t = paste0(ref_targets[1], "_", condition, "_vs_LUC_", condition)
t2 = paste0(ref_targets[2], "_", condition, "_vs_LUC_", condition)

comp = gsub("_", " ", t)
comp2 = gsub("_", " ", t2)

t = gsub("([.|()\\^{}+$*?]|\\[|\\])", "\\\\\\1", t)
t2 = gsub("([.|()\\^{}+$*?]|\\[|\\])", "\\\\\\1", t2)

f1 = list_dir[grepl(t, list_dir)]
f2 = list_dir[grepl(t2, list_dir)]

csv1 = file.path(f1, paste0("PcB_targets_", basename(f1), ".csv"))
csv2 = file.path(f2, paste0("PcB_targets_", basename(f2), ".csv"))

tab1 = read.csv(csv1)[,-1]
tab2 = read.csv(csv2)[,-1]

limits = c(min(tab1$padj, tab2$padj, na.rm = TRUE), 1) %>% rev()
limits = -log10(limits)

for(i in 1:length(targets)){
  # Load results table
  t = paste0(targets[i], "_", condition, "_vs_LUC_", condition)

  comp = gsub("_", " ", t)

  t_gsub = gsub("([.|()\\^{}+$*?]|\\[|\\])", "\\\\\\1", t)

  f1 = list_dir[grepl(t_gsub, list_dir)]

  csv1 = file.path(f1, paste0("PcB_targets_", basename(f1), ".csv"))

  tab1 = read.csv(csv1)[,-1]
  
  csv2 = file.path(f1, paste0("results_", basename(f1), ".csv"))

  tab2 = read.csv(csv2)[,-1]

  # Prepare data for volcano plot
  comparison = gsub("_", " ", t)
  comp_dir = file.path(results_dir, subfolder, t)
    if(!dir.exists(comp_dir)){
      dir.create(comp_dir)
    }

  # Make volcano plot and export
  vp = makeVolcanoPlot(tab1, lfc.threshold = lfc, ylim = limits)

  # Export volcano plot
  vp_path = file.path(comp_dir,
                      paste0("volcano_scaled_PcB_targets", t, ".pdf"))

  pdf(file = vp_path, width = 6.5, height = 5)
  print(vp)
  dev.off()
  
  # Make volcano plot including all genes and export
  vp2 = makeVolcanoPlot2(tab2, tab1, lfc.threshold = lfc, ylim = limits)

  # Export volcano plot
  vp_path = file.path(comp_dir,
                      paste0("volcano_scaled_all_genes_", t, ".pdf"))

  pdf(file = vp_path, width = 6.5, height = 5)
  print(vp2)
  dev.off()

}
```

```{r Find common genes among comparisons}
condition = "K27M+"
targets = c("CBX4.1", "CBX4.2", "BMI1.1", "BMI1.3", "EZH2.3")

subfolder = "Harvest plus Condition"

search_dir = file.path(results_dir, subfolder)
list_dir = list.dirs(search_dir)

targets_list = character()

for(i in 1:length(targets)){
  t = paste0(targets[i], "_", condition, "_vs_LUC_", condition)
  t = gsub("([.|()\\^{}+$*?]|\\[|\\])", "\\\\\\1", t)
  
  f1 = list_dir[grepl(t, list_dir)]
  targets_list = append(targets_list, f1)
}

# Create list of the results tables for each target
results_list = list()

for(i in 1:length(targets_list)){
  csv1 = file.path(targets_list[i], paste0("PcB_targets_", 
                                           basename(targets_list[i]), 
                                           ".csv"))
  tab1 = read.csv(csv1)[,-1]
  results_list[[i]] = tab1
}

names(results_list) <- targets

# Filter results to get only DE genes on each data frame
filtered_results_list <- lapply(results_list, function(df) {
  df[df$effect != "NS", ]
})

patterns = c("Gene|log2FoldChange|padj|effect")

# Append the respective name as a suffix to each column name in each data frame
for (i in seq_along(results_list)) {
  df_name <- names(results_list)[i]
  colnames(filtered_results_list[[i]]) <- paste0(colnames(filtered_results_list[[i]]), "_", df_name)
  colnames(filtered_results_list[[i]])[1] = "Gene"
  cols = colnames(filtered_results_list[[i]])[grepl(patterns, 
                                           colnames(filtered_results_list[[i]]))]
  filtered_results_list[[i]] = filtered_results_list[[i]][cols]
}

# Merge DE info for each target into a single table to get common genes
common_genes = filtered_results_list[[1]]

for(i in 2:length(filtered_results_list)){
  common_genes = merge(common_genes, filtered_results_list[[i]], by = "Gene")
}
  
# Export table of common genes
df_path = file.path(results_dir,
                    paste0("common_genes_PcB_targets.csv"))
write.csv(common_genes, file = df_path, quote = FALSE, row.names = F)

```


### MDS plots (all genotypes)
```{r MDS plot (all genotypes), fig.height=5, fig.width= 6.5}
# Subset by genotype
counts_WT = counts[, grepl("K27M+", colnames(counts))]
counts_KO = counts[, grepl("K27M_KO", colnames(counts))]

metadata_WT = filter(sample_metadata, genotype == "K27M+")
metadata_KO = filter(sample_metadata, genotype == "K27M_KO")

# Set up colors by genotype
leg1 = sample_metadata$harvest %>% factor() %>% unique()
leg1_sub = gsub("_|-", " ", leg1)
cols1 = sample_metadata$harvest %>% factor() %>% as.numeric() 
shape1 = sample_metadata$genotype %>% factor() %>% as.numeric()

# Set up labels for the plot
labels1 = sample_metadata$guide %>% factor() 

library(edgeR)

# Add extra space to right of plot area; change clipping to figure
par(mar=c(5.3, 4.3, 4.3, 8.3), xpd=TRUE)

# Plot
plotMDS(counts, 
        labels = labels1,
        main = "", 
        pch = shape1,
        col = cols1, # colors
        gene.selection = "pairwise"
)
legend("topright",
       fill = leg1, 
       legend = leg1_sub, # unique() will only show unique legend labels 
       col = unique(cols1), # one color per unique legend label
       inset=c(-0.2,0), # legend outside of plot
       title = "Harvest"
)



```

### MDS plots (by genotypes)
```{r MDS plot (by genotypes), fig.height=10, fig.width= 6.5}
# Subset by genotype
counts_df = as.data.frame(counts)
counts_WT = counts_df[, grepl("K27M\\+", colnames(counts_df))] %>% as.matrix()
counts_KO = counts_df[, grepl("K27M_KO", colnames(counts_df))] %>% as.matrix()

metadata_WT = filter(sample_metadata, genotype == "K27M+")
metadata_KO = filter(sample_metadata, genotype == "K27M_KO")

# Set up colors by harvest
leg1 = metadata_WT$harvest %>% factor() %>% unique()
leg1_sub = gsub("_|-", " ", leg1)
cols1 = metadata_WT$harvest %>% factor() %>% as.numeric() 

leg2 = metadata_KO$harvest %>% factor() %>% unique()
leg2_sub = gsub("_|-", " ", leg1)
cols2 = metadata_KO$harvest %>% factor() %>% as.numeric() 

# Set up labels for the plots
labels1 = metadata_WT$guide %>% factor() 
labels2 = metadata_KO$guide %>% factor() 

library(edgeR)

# Add extra space to right of plot area; change clipping to figure
par(mar=c(5.3, 4.3, 4.3, 8.3), xpd=TRUE, mfrow=c(2,1) )

# Plot
plotMDS(counts_WT, 
        labels = labels1,
        main = "K27M+", 
        col = cols1, # colors
        gene.selection = "pairwise", cex = 0.75
)
legend("topright",
       fill = leg1, 
       legend = leg1_sub, # unique() will only show unique legend labels 
       col = unique(cols1), # one color per unique legend label
       inset=c(-0.2,0), # legend outside of plot
       title = "Harvest"
)

par(mar=c(5.3, 4.3, 4.3, 8.3), xpd=TRUE )

# Plot
plotMDS(counts_KO, 
        labels = labels2,
        main = "K27M KO", cex = 0.75, 
        col = cols2, # colors
        gene.selection = "pairwise"
)
legend("topright",
       fill = leg2, 
       legend = leg2_sub, # unique() will only show unique legend labels 
       col = unique(cols2), # one color per unique legend label
       inset=c(-0.2,0), # legend outside of plot
       title = "Harvest"
)
```



